docs/man/rpm-dependency-generators.7.scd

```diff
@@ -0,0 +1,381 @@
+information to packages at build time. This system relies on specialized, domain
+specific executables, called _generators_, that do the heavy-lifting of figuring
+out the dependencies from the files present in the buildroot. It complements the
+manual tags, such as *Requires* or *Provides*, in *rpm-spec*(5) files.
+
+Generators are paired with logical labels, called _file attributes_, assigned to
+the packaged files by the file classifier. This ensures that every file is only
+processed by the applicable generator(s), as opposed to all generators, which
+simplifies their implementation and significantly improves the overall build
+performance.
+
+# FILE ATTRIBUTES
+
+A file attribute is a logical label that RPM gives to one or more files in the
+buildroot, based on a set of matching _rules_. It can represent a single file
+type, such as ELF, or a broader domain, such as all source files written in a
+specific programming language.
+
+These attributes are distinct from the file attributes in the filesystem and are
+only kept in memory during package builds. A single file may have an arbitrary
+number of attributes, and a single attribute may be assigned to an arbitrary
+number of files.
+
+A valid file attribute called _NAME_ is introduced with an *rpm-macrofile*(5),
+_NAME.attr_, dropped into the *%{\_fileattrsdir}* directory, that defines at
+least one of the following macros:
+
+```
+%__NAME_path
+%__NAME_magic
+%__NAME_exclude_path
+%__NAME_exclude_magic
+```
+
+The values of these macros represent the inclusion and exclusion rules. They are
+macro-expanded on use and then interpreted as extended regular expressions (RE).
+Matching is inclusive unless changed by _flags_ (see below): If the path, magic
+or both match, the file is said to _have_ the attribute in question, unless
+there is a matching exclusion rule (also inclusive by default) or a flag that
+prevents the match.
+
+The path REs are matched against the packaged path, that is, without the
+*%buildroot* prefix (for example, _/bin/bash_). The magic REs are matched
+against the result of *magic*(5) classification and possibly the *FILECLASS* tag
+stored in the package (see *--fileclass* in *rpm*(8) for details).
+
+Flags are comma-separated lists specified with the following macros:
+
+```
+%__NAME_flags
+%__NAME_exclude_flags
+```
+
+Currently, the following flags are supported (Added: 4.9.1):
+- *exeonly*: require executable bit set
+- *magic_and_path*: require both magic and path to match
+
+# GENERATORS
+
+A generator is an executable that reads filename(s) from stdin and writes
+dependencies of a single _type_ to stdout, one per line. Unlike the path REs,
+generators receive filenames with the *%buildroot* prefix since they typically
+need to look at the actual file contents on disk. This simple protocol allows
+for the generators to be implemented in any preferred language and enables them
+to use language specific libraries or tools.
```

Comment:
For man page style, the last sentence could/should perhaps be replaced with a simple "generators can be written in any programming language" or something to that tune.

Reply:
Ack, this is indeed an interesting observation (much like the one in the intro paragraph about the file attributes "justification"). It does make sense, though. Man pages shall be written in a uniform style, one that simply *states* things, as opposed to explain them, I guess.